src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php

```diff
@@ -867,4 +895,454 @@ public static function arithmeticOperation(
+            if (!$result_type) {
+                $result_type = new Type\Union([new Type\Atomic\TInt()]);
+            } else {
+                $result_type = Type::combineUnionTypes(
+                    new Type\Union([new Type\Atomic\TInt()]),
+                    $result_type
+                );
+            }
+            return;
+        }
+
+        if ($parent instanceof PhpParser\Node\Expr\BinaryOp\Mul) {
+            self::analyzeMulBetweenIntRange($right_type_part, $left_type_part, $parent, $result_type);
+            return;
+        }
+
+        if ($parent instanceof PhpParser\Node\Expr\BinaryOp\Pow) {
+            self::analyzePowBetweenIntRange($left_type_part, $right_type_part, $result_type);
+            return;
+        }
+
+        if ($parent instanceof PhpParser\Node\Expr\BinaryOp\Minus) {
+            //for Minus, we have to assume the min is the min from first range minus the max from the second
+            $min_operand1 = $left_type_part->min_bound;
+            $min_operand2 = $right_type_part->max_bound;
+            //and the max is the max from first range minus the min from the second
+            $max_operand1 = $left_type_part->max_bound;
+            $max_operand2 = $right_type_part->min_bound;
+        } else {
+            $min_operand1 = $left_type_part->min_bound;
+            $min_operand2 = $right_type_part->min_bound;
+
+            $max_operand1 = $left_type_part->max_bound;
+            $max_operand2 = $right_type_part->max_bound;
+        }
+
+        $calculated_min_type = null;
+        if ($min_operand1 !== null && $min_operand2 !== null) {
+            // when there are two valid numbers, make any operation
+            $calculated_min_type = self::arithmeticOperation(
+                $parent,
+                $min_operand1,
+                $min_operand2,
+                false
+            );
+        }
+
+        $calculated_max_type = null;
+        if ($max_operand1 !== null && $max_operand2 !== null) {
+            // when there are two valid numbers, make any operation
+            $calculated_max_type = self::arithmeticOperation(
+                $parent,
+                $max_operand1,
+                $max_operand2,
+                false
+            );
+        }
+
+        $min_value = $calculated_min_type !== null ? $calculated_min_type->getSingleIntLiteral()->value : null;
+        $max_value = $calculated_max_type !== null ? $calculated_max_type->getSingleIntLiteral()->value : null;
+
+        $new_result_type = new Type\Union([new Type\Atomic\TIntRange($min_value, $max_value)]);
+
+        if (!$result_type) {
+            $result_type = $new_result_type;
+        } else {
+            $result_type = Type::combineUnionTypes($new_result_type, $result_type);
+        }
+    }
+
+    /**
+     * @param TIntRange|TInt $left_type_part
+     * @param TIntRange|TInt $right_type_part
+     */
+    private static function analyzeOperandsBetweenIntRangeAndInt(
```

Comment:
it probably 'analyze *operation*' rather than 'analyze *operands*'

Reply:
Hmm, I named that function from the caller method that was named `analyzeNonDivOperands` before this PR and now named `analyzeOperands` but Operation seems more fitting indeed. Should I rename everything from Operands to Operation?

Reply:
Let's postpone it to some time later.